File: server/Models/BreedingEvent.cs
namespace Server.Models;

/// <summary>
/// One pairing of two animals and what came of it. Recorded from a livestock group's breeding
/// page, between two <see cref="LivestockDetail"/> animals — the individuals, not the groups,
/// because a group has no gender and a herd does not breed as a unit.
/// </summary>
public class BreedingEvent
{
    public int Id { get; set; }

    /// <summary>
    /// The group whose breeding page this was recorded from. Nullable, and only ever context:
    /// the pair may come from two different groups, and either group can be removed while the
    /// event stays worth keeping. What the event is actually about is the two animals below.
    /// </summary>
    public int? LivestockId { get; set; }

    /// <summary>The sire. Null once the animal itself is gone, which does not erase the event.</summary>
    public int? MaleAnimalId { get; set; }

    /// <summary>The dam, under the same rule as <see cref="MaleAnimalId"/>.</summary>
    public int? FemaleAnimalId { get; set; }

    /// <summary>When the pairing happened — given by the farmer, not stamped by the server, since
    /// it is routinely recorded after the fact. Doubles as the date of the
    /// <see cref="BreedingStatus.Breeding"/> stage.</summary>
    public DateOnly BreedingDate { get; set; }

    public string? Comment { get; set; }

    public BreedingStatus Status { get; set; } = BreedingStatus.Breeding;

    /// <summary>
    /// When the event reached <see cref="BreedingStatus.PregnancyConfirmed"/>. Stamped by the
    /// server the first time the status is moved there and left alone afterwards, so an event that
    /// went on to fail still says when the pregnancy had been confirmed.
    /// </summary>
    public DateOnly? PregnancyConfirmedDate { get; set; }

    /// <summary>When the event reached <see cref="BreedingStatus.Completed"/>.</summary>
    public DateOnly? CompletedDate { get; set; }

    /// <summary>When the event reached <see cref="BreedingStatus.Failed"/>.</summary>
    public DateOnly? FailedDate { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

File: web/src/types/breeding-event.ts
/**
 * How far a breeding event has got. Stored as these names, so the same strings appear in the
 * database — see server/Models/BreedingStatus.cs.
 */
export type BreedingStatus = 'Breeding' | 'PregnancyConfirmed' | 'Completed' | 'Failed';

/** In the order an event moves through them, which is the order the picker offers. */
export const BREEDING_STATUSES: readonly BreedingStatus[] = [
  'Breeding',
  'PregnancyConfirmed',
  'Completed',
  'Failed',
];

export type BreedingEvent = {
  id: number;
  /** The group whose breeding page this was recorded from — context only, and null once that
   *  group is gone. The pair below is what the event is about. */
  livestockId: number | null;
  /** The sire, as a LivestockDetail id. Null once the animal itself is gone. */
  maleAnimalId: number | null;
  /** The dam, under the same rule as maleAnimalId. */
  femaleAnimalId: number | null;
  /** ISO date (YYYY-MM-DD) the pairing happened — given, not stamped. Also the date of the
   *  Breeding stage, which is why that stage has no date of its own below. */
  breedingDate: string;
  comment: string | null;
  status: BreedingStatus;
  /** ISO dates stamped by the server as each stage is reached, and kept once stamped so an event
   *  that moved on still says when it passed through. Null for a stage never reached. */
  pregnancyConfirmedDate: string | null;
  completedDate: string | null;
  failedDate: string | null;
  /** How many animals this pairing has produced, over every result recorded against it. 0 until
   *  one is. Held on the event because a result can be recorded as a head count with no animals
   *  written down, and those leave nothing to count. */
  offspringCount: number;
  /** The group they joined, from the most recent result. */
  offspringLivestockId: number | null;
  createdAt: string;
};

export type BreedingEventInput = Omit<BreedingEvent, 'id' | 'createdAt'>;

/** What a pairing produced. The parents are not sent — the server takes them from the event. */
export type BreedingResultInput = {
  /** The group the offspring join. Same kind as the parents. */
  livestockId: number;
  quantity: number;
  gender: 'Male' | 'Female' | null;
  imagePath: string | null;
  /** ISO date (YYYY-MM-DD) they were born, or null when not recorded. */
  bornDate: string | null;
  /** Whether each animal also gets a record of its own. Off, they are only counted: the group's
   *  head count rises and nothing else is written, so there is nothing to carry the parentage,
   *  the gender or the photo either. */
  addIndividuals: boolean;
};
